tests/components/yessssms/test_notify.py

```diff
@@ -1,7 +1,147 @@
 """The tests for the notify yessssms platform."""
 import unittest
+from unittest.mock import patch, MagicMock
+
+import pytest
 import requests_mock
+
+from homeassistant.setup import async_setup_component
 import homeassistant.components.yessssms.notify as yessssms
+from homeassistant.components.yessssms.const import CONF_PROVIDER
+
+from homeassistant.const import CONF_PASSWORD, CONF_RECIPIENT, CONF_USERNAME
+
+
+@pytest.fixture(name="config")
+def config_data():
+    """Set valid config data."""
+    config = {
+        "notify": {
+            "platform": "yessssms",
+            "name": "sms",
+            CONF_USERNAME: "06641234567",
+            CONF_PASSWORD: "secretPassword",
+            CONF_RECIPIENT: "06509876543",
+            CONF_PROVIDER: "educom",
+        }
+    }
+    return config
+
+
+@pytest.fixture(name="valid_settings")
+def init_valid_settings(hass, config):
+    """Initialize component with valid settings."""
+    return async_setup_component(hass, "notify", config)
+
+
+@pytest.fixture(name="invalid_provider_settings")
+def init_invalid_provider_settings(hass, config):
+    """Set invalid provider data and initalize component."""
+    config["notify"][CONF_PROVIDER] = "FantasyMobile"  # invalid provider
+    return async_setup_component(hass, "notify", config)
+
+
+@pytest.fixture(name="invalid_login_data")
+def mock_invalid_login_data():
+    """Mock invalid login data."""
+    path = "homeassistant.components.yessssms." "notify.YesssSMS.login_data_valid"
+    with patch(path, return_value=False):
+        yield
+
+
+@pytest.fixture(name="valid_login_data")
+def mock_valid_login_data():
+    """Mock valid login data."""
+    path = "homeassistant.components.yessssms." "notify.YesssSMS.login_data_valid"
+    with patch(path, return_value=True):
+        yield
+
+
+@pytest.fixture(name="connection_error")
+def mock_connection_error():
+    """Mock a connection error."""
+    yessssms.YesssSMS.login_data_valid = MagicMock(return_value=True)
```

Comment:
Use patch here too. Pass the side_effect as keyword argument.

Reply:
When would you mock then?
patching can do the same, can't it?

Reply:
A common use case is when we need to assert something about an attribute of the return value of a patched function. Then we create a mock first and set it as return value of the attribute of the return value when patching and then we can use the mock in our tests to assert that the attribute was used correctly in the test.

When we need to build a more complex representation of the part of the code under test where we don't want the real code, or if we need to create the mock before applying the patch, we use mocks explicitly. If we just want to temporarily replace objects in the code under test we use patch. Since patch return a mock, the two can be used together to make complex scenarios.